### DISCUSSION TITLE
i18n and composition API

### DISCUSSION THREAD

**Discussion by stormcodex** [quasarframework/quasar#12248] [General - Components / Directives / etc] [answered]
How can I use i18n in setup()?
this.$t("key") or $t("key") do not work.

Uncaught TypeError: Cannot read properties of undefined (reading '$t')

**Comment by MilosPaunovic** [comment]
First step would be to go through this segment of offical docs:
[https://quasar.dev/options/app-internationalization#introduction](https://quasar.dev/options/app-internationalization#introduction)

If that doesn't guide you to a solution, you'll need to share some code with us, for someone to be able to help you.

**Comment by stormcodex** [comment]
Here is an example of what does not work. The example function is defined within setup().

`function example() {
        $q.dialog({
          title: "Test",
          message: $t("key"),
        });
}`

**Reply by MilosPaunovic** [reply]
What kind of error are you getting? Is there anything in the console?

**Comment by yusufkandemir** [comment, marked as answer]
You need to get it like: `const { t } = useI18n(...)`
See vue-i18n docs for more information: https://vue-i18n.intlify.dev/guide/advanced/composition.html#message-translation

**Reply by stormcodex** [reply]
It works. Thank you!

**Reply by danb4r** [reply]
I did this to make it work on Composition API, bot with setup() function and with script setup syntax.. but I am pretty sure there is a better way:

`import { useI18n } from 'vue-i18n';
const t = useI18n();
var message = t.t('ok_button');
`
Could the docs be updated to reflect that?..

**Comment by CurryAyam** [comment]
i add a global properties on boot file

```
export default boot(({ app }) => {
  const i18n = createI18n({
    locale: 'id',
    // fallbackLocale: 'en-US',
    messages,
  })
  app.config.globalProperties.$t = i18n.global.t

  // Set i18n instance on app
  app.use(i18n)
})
```

**Comment by brianoflondon** [comment]
Using translations in the `<script setup>` section of the Composition API. What a mess.

First you MUST change your `i18n.js` boot file and make sure `legacy: false` is in there

```js
import { boot } from 'quasar/wrappers'
import { createI18n } from 'vue-i18n'
import messages from 'src/i18n'

export default boot(({ app }) => {
  const i18n = createI18n({
    locale: 'en-US',
    globalInjection: true,
    legacy: false,
    messages
  })

  // Set i18n instance on app
  app.use(i18n)
})
```

Then in your component this should work.

```js
<script setup>
import { useI18n } from 'vue-i18n'
const t = useI18n().t

console.log(t('login'))
```
